### PR TITLE
NDP: Year change controller and update dependant tables and values

### PR DESCRIPTION
**Pull request by sorja**
[screen-recorder-tue-oct-17-2023-11-43-32.webm](https://github.com/openforis/fra-platform/assets/5508251/7006a13e-f273-4de4-af3f-26e088ff82ac)

better view of deps updating:

[screen-recorder-wed-oct-18-2023-12-31-16.webm](https://github.com/openforis/fra-platform/assets/5508251/e7f209e2-7063-4411-8b8e-de900a259898)

### CODE REVIEW

File: src/client/store/data/hooks/useOriginalDataPointYears.ts [hunk 1]
```diff
@@ -9,6 +9,6 @@ export const useOriginalDataPointYears = (): Array<{ year: string; id: number }>
 
   return Object.entries(odpData).map(([year, data]) => ({
     year,
-    id: data.total.odpId,
+    id: data.totalLandArea.odpId,
```
Comment (sorja) [inline src/client/store/data/hooks/useOriginalDataPointYears.ts:12]: fix crash when deleting

File: src/server/controller/cycleData/originalDataPoint/removeOriginalDataPoint.ts [hunk 2]
```diff
@@ -22,6 +22,7 @@ export const removeOriginalDataPoint = async (props: Props, client: BaseProtocol
   const cycleName = cycle.name
   const { countryIso } = originalDataPoint
 
+  // TODO Pass transaction client all the way down to redis table data fetching/updating
```
Comment (sorja) [inline src/server/controller/cycleData/originalDataPoint/removeOriginalDataPoint.ts]: @minotogna thoughts?

File: src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  // --- 3 Notify client about delete
+  const { name: assessmentName } = assessment.props
+  const { name: cycleName } = cycle
+  const nodeUpdateEvent = Sockets.getODPDeleteEvent({ assessmentName, cycleName, countryIso })
+  SocketServer.emit(nodeUpdateEvent, { countryIso, year: originalDataPoint.year })
+
+  // --- 4 Update dependents
+  const commonProps = { assessment, cycle, sectionName, user, notifyClient: true }
+  await updateOriginalDataPointsDependentNodes({
```
Comment (sorja) [inline src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts:54]: should we revert the support for multiple?

File: src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  const propsEvent = { countryIso, assessmentName, cycleName }
+  const nodeUpdateEvent = Sockets.getNodeValuesUpdateEvent(propsEvent)
+  const nodesUpdated: Array<NodeUpdate> = []
+
+  colNames.forEach((colName) => {
+    originalDataPointVariables.forEach(({ variableName, tableName }) => {
+      const propsValue = { assessmentName, cycleName, colName, variableName, tableName, countryIso, data }
+      const value = RecordAssessmentDatas.getNodeValue(propsValue)
+      const nodeUpdate = {
+        tableName: value.odpId ? TableNames.originalDataPointValue : tableName,
```
Comment (sorja) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts]: if the value doesn't have odpId it is not an original data point value

in this case:
this happens when updating the deps for the deleted odp

File: src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  // Update cacheCountryTable for all countries for original data points
+  await Promise.all(
+    originalDataPoints.map((originalDataPoint) => {
+      const { countryIso } = originalDataPoint
+      return DataRedisRepository.cacheCountryTable({ assessment, cycle, countryIso, tableName, force: true })
+    })
+  )
+
+  // 2. schedule dependencies update
+  const allNodes: Record<CountryIso, Array<NodeUpdate>> = {} as Record<CountryIso, Array<NodeUpdate>>
```
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts]: in practice we don't support multiple country isos in the scheduleUpdateDependencies... so we can safely assume all odps passed are the same country iso ?

File: src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    const activityLog = { target: updatedOriginalDataPoint, section: 'odp', message, countryIso, user }
+    await ActivityLogRepository.insertActivityLog({ activityLog, assessment, cycle }, t)
+
+    return updatedOriginalDataPoint
+  })
+
+  // --- 3 Notify client about delete
+  const { name: assessmentName } = assessment.props
+  const { name: cycleName } = cycle
+  const nodeUpdateEvent = Sockets.getODPDeleteEvent({ assessmentName, cycleName, countryIso })
```
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts]: can we call this odpDeleteEvent ?

File: src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  SocketServer.emit(nodeUpdateEvent, { countryIso, year: originalDataPoint.year })
+
+  // --- 4 Update dependents
+  const commonProps = { assessment, cycle, user, notifyClient: true }
```
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts]: here we notify also about the removed year and it was previously notified already. can we avoid this (see issue) ?

File: src/server/repository/assessmentCycle/originalDataPoint/updateYear.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  const { assessment, cycle, countryIso, id, targetYear } = props
+
+  const schemaName = Schemas.getNameCycle(assessment, cycle)
+
+  await client.one<OriginalDataPoint>(
+    `
+        update    ${schemaName}.original_data_point
+        set       year = $2
+        where     id = $1
+        returning *
```
Comment (minotogna) [inline src/server/repository/assessmentCycle/originalDataPoint/updateYear.ts]: this return could be avoided

File: src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  SocketServer.emit(odpDeleteEvent, { countryIso, year: originalDataPoint.year })
+
+  // --- 4 Update dependents
+  const commonProps = { assessment, cycle, user }
+  await updateOriginalDataPointsDependentNodes({
+    ...commonProps,
+    originalDataPoints: [
+      { originalDataPoint, notifyClient: false },
+      { originalDataPoint: updatedOriginalDataPoint, notifyClient: true },
+    ],
```
Comment (sorja) [inline src/server/controller/cycleData/originalDataPoint/updateOriginalDataPointYear.ts:59]: Updated the way we pass notifyClient so we can separate which updates we want to send over websocket

File: src/server/controller/cycleData/originalDataPoint/removeOriginalDataPoint.ts (hunk 2 above)
Comment (minotogna) [reply src/server/controller/cycleData/originalDataPoint/removeOriginalDataPoint.ts]: What's the problem here ?

File: src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts (hunk 4 above)
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts]: How come this is not an odp ?

File: src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    (acc, { originalDataPoint, notifyClient }) => {
+      if (notifyClient) {
+        acc.push(String(originalDataPoint.year))
+      }
+      return acc
+    },
+    []
+  )
+
+  await notifyClientUpdate({ cycle, sectionName, assessment, countryIso, colNames })
```
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts]: Can we simply pass the array of original data points and then move the logic inside notifyClient where the nodes to send are derived from odps directly ?

File: src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+  const nodeUpdates = { assessmentName, cycleName, countryIso, nodes }
+  const propsDeps = { assessment, cycle, isODP: true, nodeUpdates, user }
+  await scheduleUpdateDependencies(propsDeps)
+
+  // 3. notifies client
+  const colNames: Array<string> = originalDataPoints.reduce<Array<string>>(
+    (acc, { originalDataPoint, notifyClient }) => {
+      if (notifyClient) {
+        acc.push(String(originalDataPoint.year))
```
Comment (sorja) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts]: Move this inside of notifyClientUpdate and pass originalDataPoints to notifyClientUpdate

File: src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+  const nodesUpdated: Array<NodeUpdate> = originalDataPoints.reduce<Array<NodeUpdate>>(
+    (acc, { originalDataPoint, notifyClient }) => {
+      if (notifyClient) {
+        const colName = String(originalDataPoint.year)
+        originalDataPointVariables.forEach(({ variableName, tableName }) => {
+          const propsValue = { assessmentName, cycleName, colName, variableName, tableName, countryIso, data }
+          const value = RecordAssessmentDatas.getNodeValue(propsValue)
+          const nodeUpdate = {
+            tableName: value.odpId ? TableNames.originalDataPointValue : tableName,
```
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts]: why can't this be an odp ? maybe throw exception if  value.odpId != originalDataPoint.id ?

File: src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+  originalDataPoints: Array<{ originalDataPoint: OriginalDataPoint; notifyClient: boolean }>
+  user: User
+}
+
+export const updateOriginalDataPointsDependentNodes = async (props: Props): Promise<void> => {
+  const { assessment, cycle, sectionName, originalDataPoints: _originalDataPoints, user } = props
+  const assessmentName = assessment.props.name
+  const cycleName = cycle.name
+
+  const originalDataPoints = _originalDataPoints.filter(({ originalDataPoint }) => originalDataPoint.year)
```
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/updateOriginalDataPointsDependentNodes.ts]: can we ever pass odp with no year ? maybe throw exception if an odp doesn't have a year ?

File: src/server/repository/assessmentCycle/countryActivityLog/_common/acceptedMessages.ts [hunk 14]
```diff
@@ -13,6 +13,7 @@ export const acceptedMessages = [
   ActivityLogMessage.originalDataPointUpdateDescription,
   ActivityLogMessage.originalDataPointUpdateNationalClasses,
   ActivityLogMessage.originalDataPointUpdateOriginalData,
+  ActivityLogMessage.originalDataPointUpdateYear,
```
Comment (minotogna) [inline src/server/repository/assessmentCycle/countryActivityLog/_common/acceptedMessages.ts:16]: have we checked countrySummary view as well ?

Comment (sorja) [reply src/server/repository/assessmentCycle/countryActivityLog/_common/acceptedMessages.ts:16]: yes

https://github.com/openforis/fra-platform/blob/bf90dcdfb617bd46ecd76014bddda68eed250395/src/server/repository/assessmentCycle/countrySummary/_lastEditActivities.ts#L13

File: src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts (hunk 12 above)
Comment (sorja) [reply src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts]: It is not resulting in non-odp values anymore so this check was removed.
I suspect it was from previous iteration where we notified for both removed (non-odp value) and the updated odp.

I left the check for ODP id validness

File: src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        originalDataPointVariables.forEach(({ variableName, tableName }) => {
+          const propsValue = { assessmentName, cycleName, colName, variableName, tableName, countryIso, data }
+          const value = RecordAssessmentDatas.getNodeValue(propsValue)
+
+          if (value.odpId && value.odpId !== originalDataPoint.id)
+            throw new Error(`value.odpId ${value.odpId} is different from originalDataPoint.id ${originalDataPoint.id}`)
+
+          const nodeUpdate = { tableName: TableNames.originalDataPointValue, variableName, colName, value }
+
+          if (!Objects.isEmpty(value)) acc.push(nodeUpdate)
```
Comment (minotogna) [inline src/server/controller/cycleData/originalDataPoint/updateDependants/notifyClientUpdate.ts]: can't an odp have an empty value now and previously client didn't ?

**Review by minotogna — APPROVED** [review]

**Pull request merged**